pkg/resource/firewall/hook.go

```diff
@@ -27,3 +36,251 @@ var (
+// returned values to the Firewall spec.
+func (rm *resourceManager) addLoggingConfigToSpec(
+	ctx context.Context,
+	r *resource,
+	ko *svcapitypes.Firewall,
+) (err error) {
+	getLoggingConfigurationResponse, err := rm.sdkapi.DescribeLoggingConfigurationWithContext(ctx, rm.getLoggingConfigurationPayload(r))
+	if err != nil {
+		return err
+	}
+	ko.Spec.LoggingConfiguration = rm.setResourceLoggingConfiguration(r, getLoggingConfigurationResponse)
+	return nil
+}
+
+func customPreCompare(
+	a *resource,
+	b *resource,
+) {
+	if a.ko.Spec.LoggingConfiguration == nil && b.ko.Spec.LoggingConfiguration == nil {
+		return
+	}
+
+	if a.ko.Spec.LoggingConfiguration == nil {
+		a.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	if b.ko.Spec.LoggingConfiguration == nil {
+		b.ko.Spec.LoggingConfiguration = &svcapitypes.LoggingConfiguration{}
+	}
+
+	sort.Slice(a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *a.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+	sort.Slice(b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[:], func(i, j int) bool {
+		return *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[i].LogType < *b.ko.Spec.LoggingConfiguration.LogDestinationConfigs[j].LogType
+	})
+}
+
+func (rm *resourceManager) getLoggingConfigurationPayload(
+	r *resource,
+) *svcsdk.DescribeLoggingConfigurationInput {
+	res := &svcsdk.DescribeLoggingConfigurationInput{}
+	res.SetFirewallName(*r.ko.Spec.FirewallName)
+	return res
+}
+
+func (rm *resourceManager) newLoggingConfigurationPayload(
+	r *resource,
+) *svcsdk.UpdateLoggingConfigurationInput {
+	res := &svcsdk.UpdateLoggingConfigurationInput{}
+	res.SetFirewallName(*r.ko.Spec.FirewallName)
+	if r.ko.Spec.LoggingConfiguration != nil {
+		res.SetLoggingConfiguration(rm.newLoggingConfiguration(r))
+	} else {
+		res.SetLoggingConfiguration(&svcsdk.LoggingConfiguration{})
+	}
+	return res
+}
+
+func makeLogDestinationConfigMapKey(cfg *svcapitypes.LogDestinationConfig) string {
+	keys := make([]string, 0, len(cfg.LogDestination))
+	for k := range cfg.LogDestination {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	key := "LogDestination:"
+	for _, k := range keys {
+		v := *cfg.LogDestination[k]
+		key += fmt.Sprintf("%s:%s,", k, v)
+	}
+
+	if cfg.LogDestinationType != nil {
+		key += fmt.Sprintf("LogDestinationType:%s,", *cfg.LogDestinationType)
+	}
+	if cfg.LogType != nil {
+		key += fmt.Sprintf("LogType:%s", *cfg.LogType)
+	}
+	return key
+}
+
+func makeLogDestinationConfigMap(logDestinationConfig []*svcapitypes.LogDestinationConfig) map[string]*svcapitypes.LogDestinationConfig {
+	logDestinationConfigMap := make(map[string]*svcapitypes.LogDestinationConfig)
+
+	for _, cfg := range logDestinationConfig {
+		logDestinationConfigMap[makeLogDestinationConfigMapKey(cfg)] = cfg
+	}
+	return logDestinationConfigMap
+}
```

Comment:
Can you walk me through why this needed? looks like from API we can use JSON objects, maybe a json.Marshall could do?

Reply:
I am not sure about what you mean nu json.Marshall solution. The purpose of the map is mentioned in the below comment.